File: src/main/java/org/springframework/data/r2dbc/repository/support/DslPreparedOperation.java
package org.springframework.data.r2dbc.repository.support;

import org.springframework.data.relational.core.sql.Delete;
import org.springframework.data.relational.core.sql.Insert;
import org.springframework.data.relational.core.sql.Select;
import org.springframework.data.relational.core.sql.Update;
import org.springframework.data.relational.core.sql.render.RenderContext;
import org.springframework.data.relational.core.sql.render.SqlRenderer;
import org.springframework.r2dbc.core.PreparedOperation;
import org.springframework.r2dbc.core.binding.BindTarget;
import org.springframework.r2dbc.core.binding.Bindings;

class DslPreparedOperation<T> implements PreparedOperation<T> {
    private T source;
    private RenderContext renderContext;
    private Bindings bindings;

    public DslPreparedOperation(T source, RenderContext renderContext, Bindings bindings) {
        this.source = source;
        this.renderContext = renderContext;
        this.bindings = bindings;
    }

    public Bindings getBindings() {
        return bindings;
    }

    @Override
    public T getSource() {
        return source;
    }

    @Override
    public void bindTo(BindTarget target) {
        bindings.apply(target);
    }

    @Override
    public String toQuery() {
        SqlRenderer sqlRenderer = SqlRenderer.create(renderContext);
        if (source instanceof Select) {
            return sqlRenderer.render((Select) source);
        }
        if (source instanceof Insert) {
            return sqlRenderer.render((Insert) source);
        }
        if (source instanceof Update) {
            return sqlRenderer.render((Update) source);
        }
        if (source instanceof Delete) {
            return sqlRenderer.render((Delete) source);
        }
        throw new IllegalStateException("Cannot render " + getSource());
    }
}

File: src/main/java/org/springframework/data/r2dbc/repository/support/DslPreparedOperation.kt
package org.springframework.data.r2dbc.repository.support

import org.springframework.data.relational.core.sql.Delete
import org.springframework.data.relational.core.sql.Insert
import org.springframework.data.relational.core.sql.Select
import org.springframework.data.relational.core.sql.Update
import org.springframework.data.relational.core.sql.render.RenderContext
import org.springframework.data.relational.core.sql.render.SqlRenderer
import org.springframework.r2dbc.core.PreparedOperation
import org.springframework.r2dbc.core.binding.BindTarget
import org.springframework.r2dbc.core.binding.Bindings

class DslPreparedOperation<T>(private val source: T, private val renderContext: RenderContext, private val bindings: Bindings) :
    PreparedOperation<T> {

    override fun getSource(): T {
        return source
    }

    override fun toQuery(): String {
        val sqlRenderer = SqlRenderer.create(renderContext)
        if (source is Select) {
            return sqlRenderer.render((source as Select))
        }
        if (source is Insert) {
            return sqlRenderer.render((source as Insert))
        }
        if (source is Update) {
            return sqlRenderer.render((source as Update))
        }
        if (source is Delete) {
            return sqlRenderer.render((source as Delete))
        }
        throw IllegalStateException("Cannot render " + getSource())
    }

    override fun bindTo(to: BindTarget) {
        bindings.apply(to)
    }
}